docs/design-documents/20221028-dead-letter-queue.md

```diff
@@ -0,0 +1,144 @@
+# Dead letter queue
+
+## Introduction
+
+In Conduit the concept of a Dead letter queue (DLQ) describes the place where
+messages are sent if they are negatively acknowledged (nacked). This can happen
+for a number of reasons:
+
+- A node in the pipeline experiences an error while processing a message (e.g.
+  processor).
+- The destination connector fails to process a message and sends back a nack.
+- The pipeline is stopped forcefully causing in-flight messages to be nacked.
+
+### Goals
+
+The purpose of a DLQ is to make a pipeline resilient to errors that might
+happen while processing messages.
+
+The design for the DLQ should support the following:
+
+- Allow user to reroute nacked messages to any Conduit connector (builtin or
+  standalone).
+- A message sent to the DLQ should reflect the contents it had just before the
+  error happened. Example: if a message went through 2 processors and failed
+  while being processed by a 3rd processor, the DLQ message should reflect any
+  transforms done by the first 2 processors.
+- Ensure message order is preserved when messages are sent to the DLQ.
+  In other words - the order in which messages were produced by the source
+  connector should be preserved, even if messages are nacked in a different
+  order.
+- If a message failed to be sent to the DLQ the nack action should fail and
+  stop the pipeline.
+- Allow the user to configure the window and percentage of messages that need
+  to be nacked before the pipeline is considered broken and stops. Example: if
+  90% of the last 30 messages are nacked stop the pipeline.
+- Messages pushed to the DLQ should contain metadata describing the error and
+  the source connector.
+
+### Questions
+
+- **Should we allow the user to attach custom metadata to messages pushed to
+  the DLQ?**
```

Comment:
in what case would this be useful for the user?

Reply:
Good question. Given that all DLQ messages would get the same metadata I don't know if it would be that useful.

Reply:
if we can come up with a useful use case, then sure, we'll give the user the ability to attach to the metadata! but for now I can't think of any either.

Reply:
In case the DLQ is a connector, it's relatively easy to add some metadata through a processor.

Reply:
Right, we could add a processor in front of the DLQ connector. The problem is how we expose this functionality to the user, it might unnecessarily complicate things, so if we see no reason I'd rather skip it altogether and keep it simple.

Reply:
>  The problem is how we expose this functionality to the user

Maybe in the same way as a normal processor, i.e. a setting under the DLQ connector? That would make a DLQ connector behave like any connector, which would maybe simplify things (no special rules like "it mustn't have a processor").

Reply:
I'm thinking that this could be a "nice to have" and not necessarily something that needs to be in the scope of the first implementation. My proposal - we write it down as future work, create an issue for it, and if time is left we add it. WDYT?

Reply:
Agree 200% with that :+1: